raster-foundry/raster-foundry
# Create modal service to centralize modal state

## Overview

This PR adds a modal service that manages the state of modals across the application. This allows for some simpler flows when it comes to using modals.

The service has two main methods: `open` and `closeActiveModal`.

**`open(modalConfig, closeActive = true)`**
This is used exactly like the original call to `$uibModal.open`. Calling the method through the service allows the service to keep track of the modal. By default this closes any other active modal. Passing `false` to `closeActive` keeps any currently open modal open.

**`closeActiveModal()`**
This dismisses the active modal.

We can now leverage this service to globally close any modal on location change, which eliminates the remnant modal problems when the user hits back with a modal open, or is forced to logout when a modal is open. In addition, this eliminates a bit of duplicated code as each controller no longer needs to manually close or keep track of active modals.

### Checklist

~- [ ] Styleguide updated, if necessary~
~- [ ] Swagger specification updated, if necessary~
~- [ ] Symlinks from new migrations present or corrected for any new migrations~

### Notes

If a call to `open` passes `false` to `closeActive` and the previously active modal is not properly managed, you could end up with an orphaned modal. This could be dealt with by keeping track of previously active modals at the service level but for now seems like a premature optimization as all of our calls to `open` do not override this default param.

## Testing Instructions

 * Spot test a few of the modals within the app -> make sure they open as expected and that when you hit the back or forward buttons that the modal is dismissed
 * Make sure to test a few of the multi-stage modals (project creation is one)

Closes #2597
